<a href="https://colab.research.google.com/github/edmenciab733/ecc_qiskit_fallfest/blob/main/operaciones_cuanticas_numpy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
%%capture
!pip install qiskit
!pip install qiskit-aer

In [2]:
import numpy as np
import qiskit

# Introducción

Este notebook es una introducción al manejo de operaciones en la computación cuántica, con el apoyo de NumPy. Será útil para comprender mejor cómo se manejan los estados cuánticos dentro del álgebra lineal, evitando la inicialmente confusa notación de Dirac.  El objetivo principal es proporcionar una comprensión sólida de cómo, desde la perspectiva de un programador y con un enfoque práctico, las operaciones cuánticas pueden ser representadas y ejecutadas como un conjunto de operaciones matriciales y vectoriales, haciendo uso de arreglos (o vectores, dependiendo del contexto específico) en un entorno de programación.

## Funciónes de ayuda

In [3]:
def separar_amplitudes(estado):
    estado = estado.reshape(2)
    num_q = int(np.log2(len(estado))) # saber a cuantos qubits corresponde su máxima longitud
    for i, amplitude in enumerate(estado):
        temp_b = format(i, "b") # convertir en binario el la posición del vector para poder
        temp_b = temp_b.zfill(num_q) # rellenar con zeros a la izquierda
        print("{} |{}>".format(amplitude, temp_b)) # formatear el resultado

# Bases Computacional , Base Hadamard, y Base  Y

---

**Base computacional**

- $ |0\rangle $
- $ |1\rangle $

**Operador de Hadamard y transformación a la Base Hadamard**

El operador de Hadamard es una operación unitaria que actúa sobre un único qubit y tiene una variedad de aplicaciones importantes, desde la creación de superposiciones hasta la implementación de algoritmos cuánticos.


- $ H = \frac{1}{\sqrt{2}}\begin{pmatrix} 1 & 1 \\ 1 & -1 \end{pmatrix} $  


- $ |+\rangle $ = $H  |0\rangle $ = $ \frac{|0\rangle + |1\rangle}{\sqrt{2}} $


- $ |-\rangle $= $H  |1\rangle $ = $ \frac{|0\rangle - |1\rangle}{\sqrt{2}} $

**Base Y **

- $ |i\rangle $
- $ |-i\rangle $

## Operadores con numpy

In [4]:
ket_0 = np.array([[1], [0]])
ket_1 = np.array([[0], [1]])

In [5]:
separar_amplitudes(ket_0)

1 |0>
0 |1>


In [6]:
separar_amplitudes(ket_1)

0 |0>
1 |1>


In [7]:
H = 1/np.sqrt(2) * np.array([[1, 1], [1, -1]])

In [8]:
ket_p  = H.dot(ket_0)

In [9]:
separar_amplitudes(ket_p)

0.7071067811865475 |0>
0.7071067811865475 |1>


In [10]:
ket_m = H.dot(ket_1)

In [11]:
ket_ip =  1/np.sqrt(2) * np.array([[1], [1j]])

In [12]:
ket_im = 1/np.sqrt(2) * np.array([[1], [-1j]])

## Producto Interno
El producto intero, también conocido como producto escalar, es una operación matemática que toma dos vectores en este espacio y devuelve un número complejo. El producto interno se denota como:  $ \langle v|u\rangle$. Un requísito es aplicar el produnto interno es que el número de columnas del primer vector debe coincidir con el número de filas.

Ejemplo con: $ \langle0|0\rangle $

$\begin{pmatrix} 1 \\ 0  \end{pmatrix} \times \begin{pmatrix} 1 & 0  \end{pmatrix} $  



Representación numpy de un Vector Columna con un Vector Fila

In [13]:
ket_0_tc = np.conj(ket_0.T)
ket_0.shape, ket_0_tc.shape

((2, 1), (1, 2))

Producto Interno entre $ \langle 0 | - \rangle $

In [14]:
np.vdot(ket_0_tc,  ket_m)

0.7071067811865475

Producto Interno entre $ \langle 0 | 0 \rangle $

In [15]:
np.vdot(ket_0_tc, ket_0)

1

Producto Interno entre $ \langle 0 | 1 \rangle $

In [16]:
ket_0_tc = np.conj(ket_0.T)
np.vdot(ket_0_tc, ket_1)

0

El detalle: El producto interno puede darnos un una idea de que tan similares son los estados de dos qubits(Mientras más cerca de 1, más similiares son, podría servir de una medida de correlación). Pruebas con: $ \langle 0|0 \rangle = 1 $ ó $ \langle0|1\rangle = 0 $  ó  $ \langle0|-\rangle = 0.707 $

# Relaciones entre qubits

Las relaciones entre qubits se realizan por medio producto de kronecker. Esto permite trabajar con muchos qubits sin que afecten unos con otros directamente.

$|00\rangle = |0\rangle \otimes |0\rangle $

$  |01\rangle = |0\rangle \otimes |1\rangle $

$|10\rangle = |1\rangle \otimes |0\rangle $

$|11\rangle = |1\rangle \otimes|1\rangle $




In [17]:
ket_00 = np.kron(ket_0, ket_0)
ket_01 = np.kron(ket_0, ket_1)
ket_10 = np.kron(ket_1, ket_0)
ket_11 = np.kron(ket_1, ket_1)

# Operadores(Puertas) de Pauli, combinaciones y operaciones sobre 1 qubit

Los operadores de Pauli constituyen una serie de matrices que tiene efecto sobre qubits individuales.

La puerta $ X $ es análoga al bit clasico actua sobre el eje $ x $ y su efecto es rotar el eje mencionado un valor $ \pi $

In [18]:
X = np.array([[0,1], [1, 0]])

In [19]:
Z = np.array([[1,0], [0, -1]])

In [20]:
Y = np.array([[0, -1j], [1j, 0]])

In [21]:
I = np.array([[1,0], [0,1]])

In [22]:
np.dot(X, ket_1)

array([[1],
       [0]])

Para combinar dos o n operadores cuanticos que terminaran a afectando a 2 o $ n $ qubits tambien se realiza el producto de kronecker entre los operados por ejemplo para dos puertas $ XX $ sería:

$ XX = X \otimes X $

El <a href="https://repositoriodigital.uns.edu.ar/handle/123456789/5841#:~:text=El%20producto%20de%20Kronecker%20se,el%20c%C3%A1lculo%20de%20valores%20singulares." > producto de kronocker </a>garantiza que los qubits se mantengan independiente. No es una operación sobre dos qubits

In [23]:
XX = np.kron(X, X)
XX

array([[0, 0, 0, 1],
       [0, 0, 1, 0],
       [0, 1, 0, 0],
       [1, 0, 0, 0]])

Ejemplo si queres afectar a aplicar al par de de qubits  $|00\rangle$ con la puerta $ X$ aplicado a ambos qubits. La operación es cuanto sigue:

$ XX |00\rangle $


In [24]:
np.dot(XX, ket_00)

array([[0],
       [0],
       [0],
       [1]])

En la notación de bra-ket, comúnmente utilizada en mecánica cuántica, el producto tensorial de operadores lineales se puede definir de la siguiente manera. Consideremos dos espacios vectoriales, $V$ y $U$ con vectores $ |v\rangle $ en $V$ y $ |u\rangle $ en $U$. El producto tensorial de los operadores  $ A y  B $ denotado por $ A \otimes B $ actúa sobre el espacio producto tensorial de $ U \otimes V $ por tanto se puede decir que tambien actua sobre $ u$  y $ v $






$ (A \otimes B ) ( |v\rangle \otimes |u\rangle ) =  (A \otimes |v\rangle ) (B \otimes |u\rangle)$

Ejemplo

$ ZX|10\rangle = Z|1\rangle X|0\rangle $

In [25]:
ZX = np.kron(Z, X)
ket1ket0 = np.kron(ket_1, ket_0)
ZXket1ket0 = ZX.dot(ket1ket0)
ZXket1ket0

array([[ 0],
       [ 0],
       [ 0],
       [-1]])

In [26]:
Zket1 = Z.dot(ket_1)
Xket0 = X.dot(ket_0)
Zket1Xket0 = np.kron(Zket1,Xket0 )
Zket1Xket0

array([[ 0],
       [ 0],
       [ 0],
       [-1]])

# Puertas RX, RY, RZ o Puertas de desplazamiento de fase.

Las puertas de desplazamiento de fase, o puertas de cambio de fase, en la computación cuántica se utilizan para modificar la fase de un estado cuántico sin cambiar sus probabilidades de medición. Estas puertas son esenciales para controlar y manipular las fases relativas de los estados cuánticos, lo que es crucial para muchos algoritmos cuánticos.


Las puertas desplazamientos no cambia la probabilidad de medir un qubit de manera individual, pero utiliza la propiedad de la interferencia cuantica para incidir en la medida generar del sistema, en el cual se combinan amplitudes de manera constructiva o destructiva. Normalmente inciden en la medida de aquellos estados entrelazados.

$RX(\theta) =  \begin{pmatrix} \cos(\frac{\phi}{2}) & -i * \sin(\frac{\phi}{2}) \\  -i * \sin(\frac{\phi}{2}) & \cos(\frac{\phi}{2}) \end{pmatrix}$

----------
$RZ(\theta) =  \begin{pmatrix} e^{-i\theta/2} & 0 \\
0 & e^{i\theta/2} \end{pmatrix}$

----------

$RY(\theta) =  \begin{pmatrix} \cos(\frac{\phi}{2}) & -\sin(\frac{\phi}{2})  \\
\sin(\frac{\phi}{2})  & \cos(\frac{\phi}{2}) \end{pmatrix}$


In [27]:
def get_matrix_rotacion(puerta, theta=np.pi):
  if puerta == "x":
    return np.array([[np.cos(theta/2), -1j * np.sin(theta/2)], [-1j * np.sin(theta/2), np.cos(theta/2) ]])
  if puerta == "z":
    return np.array([[np.exp(-1j*theta/2), 0], [0 , np.exp(1j*theta/2)  ]])
  if puerta == "y":
    return np.array([[np.cos(theta/2), -1j* np.sin(theta/2)], [np.sin(theta/2) ,np.cos(theta/2)  ]])

In [28]:
rx = get_matrix_rotacion("x")
rz = get_matrix_rotacion("z")
ry = get_matrix_rotacion("y")
aux = rz.dot(rx.dot(ket_0))

### Ejemplo  de una Puerta $RX(\pi)$ con numpy

In [29]:
print(np.round(rx, 5))

[[0.+0.j 0.-1.j]
 [0.-1.j 0.+0.j]]


### Ejemplo con qiskit de una Puerta $RX(\pi)$


In [30]:
from qiskit import QuantumCircuit
from qiskit.visualization import array_to_latex
from qiskit.quantum_info import Operator
qc = QuantumCircuit(1,1)
qc.rx(np.pi, 0)
operator = Operator(qc)
display(array_to_latex(operator, prefix="\\text{Circuito como matriz unitaria} = "))

<IPython.core.display.Latex object>

## Operación de desplazamiento de fase. Efecto sobre el $ |0\rangle $

La puerta $ RX(\pi) $, aplica una rotación de 180 grados y un cambio en una fase global del sistema. Eso a diferencia de la puerta $X$ que hace el mismo efecto, pero sin cambio fase.

In [31]:
np.round(rx.dot(ket_0), 5)

array([[0.+0.j],
       [0.-1.j]])

Si queremos tener el mismo efecto de la puerta $X$. Se puede hacer aplicando con la puerta $RZ(\pi) RX(\pi) |0\rangle $

In [32]:
np.round(rz.dot(rx.dot(ket_0)), 5)

array([[0.-0.j],
       [1.-0.j]])

# Puertas U

Es la operación que permite representar cualquier puerta cuantica. Sus parametros son los angulos que representan la esfera de bloch. Las puertas $ U $ cubren por ejemplo, las puertas de Pauli, la puerta Hadamard y las puertas de Desplazamiento de fase. Para resolver esto necesitamos entender la <a href="https://www.youtube.com/watch?v=eaP_rbGWUW4" target="_blank">exponencial compleja </a>

$U(\theta ,\phi, \lambda ) =  \begin{pmatrix} \cos(\frac{\theta}{2}) & -e^{i\lambda}\sin(\frac{\theta}{2})\\
 e^{i\phi}\sin(\frac{\theta}{2}) & e^{i(\phi + \lambda)}\cos(\frac{\theta}{2}) \end{pmatrix}$

In [92]:
def get_matrix_u(theta = np.pi, psi =  np.pi,  lamb =  np.pi ):
  return np.array([[np.cos(theta/2), -np.exp(1j*lamb) * np.sin(theta / 2) ], [np.exp(1j*psi) * np.sin(theta / 2), np.exp(1j*(psi + lamb)) * np.cos(theta / 2) ]])
  #return np.array([[np.cos(theta/2),  np.exp(1j*psi) * np.sin(theta / 2) ], [-np.exp(1j*lamb) * np.sin(theta / 2) , np.exp(1j*(psi + lamb)) * np.cos(theta / 2) ]])

### Pruebas individuales de operaciones

In [65]:
lamb = np.pi
theta = np.pi
psi = 0
np.exp(1j*(psi + lamb))

(-1+1.2246467991473532e-16j)

In [57]:
theta/2

1.5707963267948966

In [35]:
1j*lamb

3.141592653589793j

In [36]:
np.round(-np.exp(1j*lamb) * np.sin(theta / 2) , 5)

(1-0j)

### Puerta X representada como una puerta U
Operación del  $ U(\pi, 0, \pi ) |0\rangle  = X |0\rangle $

In [87]:
x_u = get_matrix_u(theta = np.pi, psi =0, lamb =np.pi)

In [83]:
print(x_u.real)

[[ 6.123234e-17  1.000000e+00]
 [ 1.000000e+00 -6.123234e-17]]


In [88]:
print( np.round(x_u.dot(ket_1), 5)   )

[[ 1.-0.j]
 [-0.+0.j]]


### Puerta Z representada como una puerta U

Operación del $ U(0, 0, \pi) |1\rangle = Z |1\rangle $

In [89]:
z_u = get_matrix_u(theta = 0, psi = 0, lamb =  np.pi)
print( np.round(z_u.dot(ket_1), 5)   )

[[ 0.+0.j]
 [-1.+0.j]]


### Puerta H representada como una puerta U

La puerta de Hadamard es una puerta cuántica fundamental que pone un qubit en una superposición equitativa de  $ |0\rangle $ ó $ |1\rangle $, no es una rotación estándar alrededor de un solo eje en la esfera de Bloch, sino que es una combinación de rotaciones y por lo tanto no se puede describir con una única puerta unitaria genérica si seguimos la forma estándar de esta puerta. Sin embargo, si permitimos una fase global adicional, que no tiene efecto en las mediciones cuánticas, podríamos expresar una transformación equivalente a la de Hadamard.

Para lograr el cambio de fase, podemos multiplica a la derecha como cambio de fase global:


  \begin{pmatrix} 1 & 0 \\ 0 & i \end{pmatrix}


Para que esta matriz sea igual a la matriz de Hadamard, los elementos deben cumplir con las siguientes condiciones:

$ \cos(\frac{\theta}{2}) = \frac{1}{\sqrt{2}} $

$ -e^{i\lambda}\sin(\frac{\theta}{2})  =  \frac{1}{\sqrt{2}} $

$  e^{i\phi}\sin(\frac{\theta}{2}) = \frac{1}{\sqrt{2}} $

$   e^{i(\phi + \lambda)}\cos(\frac{\theta}{2})  = -\frac{1}{\sqrt{2}} $


Para ello, $ \theta = \frac{\pi}{2} $


In [107]:
h_u = get_matrix_u(theta=np.pi/2, psi=0, lamb =np.pi)
h_u = h_u.dot(np.array([[1, 0], [0, 1j]]))
print(h_u)

[[ 7.07106781e-01+0.j          8.65956056e-17+0.70710678j]
 [ 7.07106781e-01+0.j         -8.65956056e-17-0.70710678j]]


In [108]:
print( np.round(h_u.dot(ket_1), 5)   )

[[ 0.+0.70711j]
 [-0.-0.70711j]]


# Puertas para dos qubits

## Puerta CNOT

## Puerta SWAP

## Puerta U Controlada